xdsl/irdl.py

```diff
@@ -476,49 +504,84 @@ def from_pyrdl(pyrdl_def: type[_OpT]) -> OpDef:
         op_def = OpDef(clsdict["name"])
         for field_name, field_type in get_type_hints(
                 pyrdl_def, include_extras=True).items():
-            origin = get_origin(field_type)
-            if origin != Annotated:
+
+            if field_name in get_type_hints(Operation).keys():
                 continue
-            args = get_args(field_type)
-
-            fail = False
-            if len(args) == 3:
-                if args[1] is OperandDef:
-                    op_def.operands.append((field_name, OperandDef(args[-1])))
-                elif args[1] is VarOperandDef:
-                    op_def.operands.append(
-                        (field_name, VarOperandDef(args[-1])))
-                elif args[1] is OptOperandDef:
-                    op_def.operands.append(
-                        (field_name, OptOperandDef(args[-1])))
-                elif args[1] is VarResultDef:
-                    op_def.results.append((field_name, VarResultDef(args[-1])))
-                elif args[1] is OptResultDef:
-                    op_def.results.append((field_name, OptResultDef(args[-1])))
+
+            origin = get_origin(field_type)
+            args: tuple[Any, ...]
+            if origin is None:
+                args = (field_type, )
+            elif origin == Annotated:
+                args = get_args(field_type)
+            else:
+                args = (field_type, )
+            args = cast(tuple[Any, ...], args)
+
+            # Get attribute constraints from a list of pyrdl constraints
+            def get_constraint(
+                    pyrdl_constrs: tuple[Any, ...]) -> AttrConstraint:
+                constraints = [
+                    irdl_to_attr_constraint(pyrdl_constr)
+                    for pyrdl_constr in pyrdl_constrs
+                    if not isinstance(pyrdl_constr, IRDLAnnotations)
+                ]
+                if len(constraints) == 0:
+                    return AnyAttr()
+                if len(constraints) == 1:
+                    return constraints[0]
+                return AllOf(constraints)
+
+            if args[0] == Operand:
+                constraint = get_constraint(args[1:])
+                op_def.operands.append((field_name, OperandDef(constraint)))
+            elif args[0] == list[Operand]:
+                constraint = get_constraint(args[1:])
+                op_def.operands.append((field_name, VarOperandDef(constraint)))
+            elif args[0] == (Operand | None):
+                constraint = get_constraint(args[1:])
+                op_def.operands.append((field_name, OptOperandDef(constraint)))
+
```

Comment:
drop space?

Reply:
Actually add a comment or something to visually separate better than an empty line?